src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_LockStatement.cs

```diff
@@ -37,6 +40,45 @@ public override BoundNode VisitLockStatement(BoundLockStatement node)
                     argumentType); //need to have a non-null type here for TempHelpers.StoreToTemp.
             }
 
+            if (argumentType.IsWellKnownTypeLock())
+            {
+                if (LockBinder.TryFindLockTypeInfo(argumentType, _diagnostics, rewrittenArgument.Syntax) is not { } lockTypeInfo)
```

Comment:
> is not { } lockTypeInfo [](http://example.com/codeflow?start=104&length=24)

I understand the desire to use pattern matting, but I personally find this code unnecessary obscure.
For example, the following would look much more clear in my opinion:
```
LockTypeInfo? lockTypeInfo = LockBinder.TryFindLockTypeInfo(argumentType, _diagnostics, rewrittenArgument.Syntax);

                if (lockTypeInfo is null)
                {
                    Debug.Fail("We should have reported an error during binding if lock type info cannot be found.");
                    return node.Update(rewrittenArgument, rewrittenBody).WithHasErrors();
                }
```

Reply:
Not sure if your comment still applies when I change `TryFindLockTypeInfo` return type to a tuple (then it would have to be `var lockTypeInfo = ...` and `lockTypeInfo.Value` wherever used below). So then this "pattern deconstruction" seems clearer to me.